### PR TITLE
adding phase dependence to prop. effects

### PR DESCRIPTION
**Pull request by jpierel14**
@benjaminrose PR to add phase dependence to propagation effects. I don't think this should cause any issues (I've tested phase-dependent, wavelength-dependent, and phase+wavelength-dependent effects), but happy to discuss.

### CODE REVIEW

**Comment by benjaminrose** [comment]
It looks like there are some errors in the testing. I know some are on SNCosmo's side but @jpierel14 does the p39 tests pass locally?

**Comment by benjaminrose** [comment]
Justin and I talked a bit offline. We noted that making phase be the first parameter (though logical if we were just starting) breaks backwards compatibility (see failing tests). Likely it is best to make phase an optional argument rather than a required positional argument.

**Comment by benjaminrose** [comment]
@jpierel14 do you know why `propagate` is defined so many times?

To get this merged, I think we should
- [ ] add a test
- [ ] shift phase to an optional parameter to keep backwards compatibility (`def propagate(self, wave, flux, phase=None):`)
- [ ] also see if there is a need to refactor how `propagate` is defined. We might fix that here, or it might be its own issue.

**Comment by benjaminrose** [comment]
The code style error is an easy fix, and I will do that. Let me also review the changes to look for maintainability problems. Other than that, it looks like we are on the right path to adding this new feature.

File: sncosmo/models.py
Comment (benjaminrose) [inline sncosmo/models.py:1960]: Is it possible to have L1956-1960 contain the `np.nan` default values, rather than adding them to all previously built models? I think this might be
```python
try:
    return self._maxphase
except NameError:
    return np.nan
```

This might also help with  backwards compatibility. If this works, then we can remove the changes to `CCM89Dust`, `OD94Dust`, and L1570-1573.

Comment (jpierel14) [reply sncosmo/models.py:1960]: Sure I think that would work let me do that.

**Comment by jpierel14** [comment]
@benjaminrose Okay new try

**Comment by jpierel14** [comment]
Look at all those check marks

**Comment by jpierel14** [comment]
>

**Comment by benjaminrose** [comment]
I have a few documentation additions. I was going to do them myself, but I con't have push access to your repo. I'll just push them to main after I merge this in.

**Review by benjaminrose — APPROVED** [review]

**Comment by jpierel14** [comment]
Sounds good, thanks a lot!

**Pull request merged**
